File: web/src/types/asset.ts
export type AssetType = 'web' | 'api' | 'mobile' | 'infra';
export type RiskLevel = 'critical' | 'high' | 'medium' | 'low' | 'info';

export interface CertInfo {
  subject: string;
  issuer: string;
  expires_at: string;
}

export interface VulnCount {
  critical: number;
  high: number;
  medium: number;
  low: number;
}

export interface Asset {
  id: string;
  domain: string;
  ip_address: string;
  asset_type: AssetType;
  confidence: number;
  risk_level: RiskLevel;
  icp_verified: boolean;
  waf_detected: string | null;
  tech_stack: string[];
  open_ports: number[];
  cert_info: CertInfo | null;
  vuln_count: VulnCount;
  interface_count: number;
  discovered_at: string;
}

// Extended asset types for column definitions
export interface IPAsset extends Asset {
  fingerprints: string[];
  related_domains: string[];
  related_units: string[];
  source: string;
  is_cdn: boolean;
}

export interface DomainAsset extends Asset {
  resolve_records: Array<{ ip: string; port: number }>;
  related_units: string[];
  related_certs: Array<{ subject: string }>;
  icp_number: string;
  cloud_provider: string;
  is_cdn_wildcard: boolean;
}

export interface WebAsset extends Asset {
  url: string;
  title: string;
  icon: string;
  screenshot: string;
  related_units: string[];
}

export interface AppAsset extends Asset {
  name: string;
  related_units: string[];
  version: string;
  download_link: string;
}

export interface MiniProgramAsset extends Asset {
  name: string;
  related_units: string[];
  platform: string;
  access_link: string;
  is_internal: boolean;
  notes: string;
}

export interface AssetListParams {
  risk?: RiskLevel;
  confirmed?: boolean;
  asset_type?: AssetType;
  icp_verified?: boolean;
  has_waf?: boolean;
  tech_stack?: string;
}

export interface UpdateAssetRequest {
  confirmed?: boolean;
  risk_level?: RiskLevel;
  notes?: string;
}

// ── 统一资产视图 (S1 资产多表 + 类型专属字段) ─────────────
// 对齐后端 TypedAssetItem (GET /v1/assets)

export type TypedAssetType = 'ip' | 'domain' | 'web' | 'app' | 'miniprogram' | 'certificate';

export interface TypedAsset {
  id: string;
  asset_type: TypedAssetType;
  domain: string | null;
  ip_address: string | null;
  port: number | null;
  risk_level: string;
  confidence: number;
  confirmed: boolean;
  company_id: string | null;
  company_name: string | null;
  tech_stack: string[];
  icp_entity: string | null;
  waf_type: string | null;
  status: string;
  source: string | null;
  notes: string | null;
  discovered_at: string | null;
  vuln_count: VulnCount;
  type_specific: Record<string, unknown>;
}

export interface TypedAssetListResponse {
  data: TypedAsset[];
  pagination: { page: number; page_size: number; total: number; total_pages: number };
}

export interface TypedAssetListParams {
  asset_type?: TypedAssetType;
  company_id?: string;
  risk?: string;
  confirmed?: boolean;
  q?: string;
  page?: number;
  page_size?: number;
}

export interface AssetFull {
  id: string;
  asset_type: string;
  domain: string | null;
  ip_address: string | null;
  port: number | null;
  risk_level: string;
  confirmed: boolean;
  company_id: string | null;
  company_name: string | null;
  tech_stack: string[];
  icp_entity: string | null;
  waf_type: string | null;
  open_ports: number[];
  cert_info: Record<string, unknown>;
  notes: string | null;
  status: string;
  value_score: number | null;
  discovered_at: string | null;
  type_specific: Record<string, unknown>;
  vulnerabilities: Array<{ id: string; title: string | null; severity: string; vuln_type: string | null; status: string | null }>;
}

File: src/api/schemas/asset.py
"""Pydantic schemas for asset API endpoints (section 9.4)."""

from __future__ import annotations

from pydantic import BaseModel, Field


# --- Request schemas ---


class DiscoveryStartRequest(BaseModel):
    dsl_queries: list[dict[str, str]] = Field(
        ...,
        min_length=1,
        description="List of {platform, query} DSL queries to search",
    )
    company_name: str | None = None
    target_domains: list[str] | None = None
    target_icp_entity: str | None = None
    target_ip_ranges: list[str] | None = None


class AssetUpdateRequest(BaseModel):
    confirmed: bool | None = None
    risk_level: str | None = Field(default=None, pattern="^(critical|high|medium|low|info)$")
    notes: str | None = None


# --- Response schemas ---


class CertInfoItem(BaseModel):
    subject: str = ""
    issuer: str = ""
    expires_at: str | None = None


class AssetItem(BaseModel):
    id: str
    domain: str | None = None
    ip_address: str | None = None
    port: int | None = None
    protocol: str = "https"
    asset_type: str = "web"
    confidence: float = 0.0
    risk_level: str = "info"
    icp_verified: bool = False
    icp_entity: str | None = None
    waf_detected: str | None = None
    tech_stack: list[str] = Field(default_factory=list)
    open_ports: list[int] = Field(default_factory=list)
    cert_info: CertInfoItem = Field(default_factory=CertInfoItem)
    screenshot_path: str | None = None
    confirmed: bool = False
    notes: str | None = None
    discovered_at: str | None = None
    # 漏洞计数 (前端 Assets 列表/详情均依赖; 默认全 0 结构避免 undefined)
    vuln_count: dict[str, int] = Field(default_factory=lambda: {"critical": 0, "high": 0, "medium": 0, "low": 0})
    interface_count: int = 0


class AssetListResponse(BaseModel):
    data: list[AssetItem] = Field(default_factory=list)
    pagination: dict = Field(default_factory=dict)


# ── 统一资产视图 (S1 资产多表 + 类型专属字段) ─────────────
class TypedAssetItem(BaseModel):
    """资产统一视图: 公共字段 + 类型专属字段 (type_specific).

    type_specific 内容随 asset_type 变化:
      ip:          {ip_address, is_cdn, asn, region, isp, open_ports}
      domain:      {domain, icp_license, icp_entity, whois_info, registrant, expires_at}
      web:         {url, title, icon, screenshot}
      app:         {name, package_name, platform, version, download_source}
      miniprogram: {name, app_id, platform, subject_entity, category}
      certificate: {common_name, issuer, serial_number, expires_at, is_expired}
    """
    id: str
    asset_type: str
    domain: str | None = None
    ip_address: str | None = None
    port: int | None = None
    risk_level: str = "info"
    confidence: float = 0.0
    confirmed: bool = False
    company_id: str | None = None
    company_name: str | None = None
    tech_stack: list[str] = Field(default_factory=list)
    icp_entity: str | None = None
    waf_type: str | None = None
    status: str = "discovered"
    source: str | None = None  # 来源: manual/icp/dns/smart_link
    notes: str | None = None
    discovered_at: str | None = None
    vuln_count: dict[str, int] = Field(default_factory=lambda: {"critical": 0, "high": 0, "medium": 0, "low": 0})
    type_specific: dict = Field(default_factory=dict, description="类型专属字段")


class TypedAssetListResponse(BaseModel):
    data: list[TypedAssetItem] = Field(default_factory=list)
    pagination: dict = Field(default_factory=dict)


class AssetDetailResponse(AssetItem):
    """资产详情 (与 AssetItem 一致, 保留以兼容现有引用)."""


class DiscoveryStartResponse(BaseModel):
    task_id: str
    status: str
    total_searched: int = 0
    total_candidates: int = 0
    total_confirmed: int = 0
    total_enriched: int = 0
    by_asset_type: dict[str, int] = Field(default_factory=dict)
    avg_confidence: float = 0.0
    errors: list[str] = Field(default_factory=list)


class DiscoveryStatusResponse(BaseModel):
    task_id: str
    status: str
    total_assets: int = 0
    total_confirmed: int = 0
    by_asset_type: dict[str, int] = Field(default_factory=dict)
    avg_confidence: float = 0.0
